main.cpp

```diff
@@ -19,15 +19,29 @@ using namespace clang::tooling;
 
 class CastCallBack : public MatchFinder::MatchCallback {
 public:
-    CastCallBack(Rewriter& rewriter) {
-        // Your code goes here
-    };
-
+    CastCallBack(Rewriter& rewriter): _rewriter(rewriter) {};
     void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
+        const auto *castExp = Result.Nodes.getNodeAs<CStyleCastExpr>("cast");
+        auto replaseRange = CharSourceRange::getCharRange(castExp->getLParenLoc(),
+                             castExp->getSubExprAsWritten()->getBeginLoc());
+        auto &srcManager = *Result.SourceManager;
+        auto type = Lexer::getSourceText(CharSourceRange::getTokenRange(
+                                              castExp->getLParenLoc().getLocWithOffset(1),
+                                              castExp->getRParenLoc().getLocWithOffset(-1)),
+                                              srcManager, LangOptions());
+        const auto *expr = castExp->getSubExprAsWritten()->IgnoreImpCasts();

```

Comment:
Please describe why you are using here `getSubExprAsWritten()` and `IgnoreImpCasts()`.
What would be the difference with the use of `getSubExprAsWritten()->IgnoreImpCasts()`, `getSubExpr()->IgnoreImpCasts()` and `getSubExprAsWritten()`?

Reply:
1) I am using getSubExprAsWritten() because I want to get subexpression as it was written in the source code, looking through any implicit casts or other intermediate nodes.
IgnoreImpCasts() - ignore any implicit casts which might surround this expression.
2) There will be no difference in our case.

Reply:
I still don't understand the need to use `getSubExprAsWritten()` and `IgnoreImpCasts()`. needs to be explained with examples.

Reply:
```
llvm::outs() << "getSubExprAsWritten()->IgnoreImpCasts(): " << castNode->getSubExprAsWritten()->IgnoreImpCasts()->getStmtClassName();
llvm::outs() << "getSubExpr()->ignoreImpCasts(): " << castNode->getSubExpr()->IgnoreImpCasts()->getStmtClassName();
llvm::outs() << "getSubExprAsWritten(): " << castNode->getSubExprAsWritten()->getStmtClassName();
```
Вывод
```
getSubExprAsWritten()->IgnoreImpCasts(): DeclRefExpr
getSubExpr()->ignoreImpCasts(): DeclRefExpr
getSubExprAsWritten(): DeclRefExpr
```
Видно, что вывод везде один и тот же.

Reply:
> I still don't understand the need to use `getSubExprAsWritten()` and `IgnoreImpCasts()`. needs to be explained with examples.

В данном случае нет необходимости, так как результат такой же

Reply:
Вставив данный код:
```
llvm::outs() « "getSubExprAsWritten()->IgnoreImpCasts(): " « castExp->getSubExprAsWritten()->IgnoreImpCasts()->getStmtClassName();
llvm::outs() « "getSubExpr()->ignoreImpCasts(): " « castExp->getSubExpr()->IgnoreImpCasts()->getStmtClassName();
llvm::outs() « "getSubExprAsWritten(): " « castExp->getSubExprAsWritten()->getStmtClassName();
```

Reply:
> I still don't understand the need to use `getSubExprAsWritten()` and `IgnoreImpCasts()`. needs to be explained with examples.

я использовал ignoreImpCasts() чтобы точно убрать все неявные приведения, но получилось, что это было не обязательно.